[package] fix mangling issues with TorchScript

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54915 [package] fix mangling issues with TorchScript**

TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

suo committed:
[package] fix mangling issues with TorchScript

TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

[ghstack-poisoned]

SplitInfinity commented on test/package/test_model.py:
```diff
@@ -187,7 +187,18 @@ def test_script_resnet(self):
+        # Model should script successfully.
+        scripted = torch.jit.script(loaded)
+
+        # Scrtiped model should save and load successfully.
```
```suggestion
        # Scripted model should save and load successfully.
```

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -926,6 +927,11 @@ def _qualified_name(obj):
+    # torch.package and TorchScript have separate mangling schemes to avoid
+    # name collisions from multiple packages. To avoid them interfering with
+    # each other, remove the package mangling here.
+    module_name = package_mangling.demangle(module_name)
```
I'm assuming this is a no-op when `module_name` has not been mangled by package?

suo replied:
correct

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]

suo committed:
Update on "[package] fix mangling issues with TorchScript"


TorchScript and torch.package have different mangling schemes. To avoid
them interfering with each other, we should undo the torch.package
mangling before processing anything with TorchScript (since TS
independently makes sure that no names collide).

Differential Revision: [D27410472](https://our.internmc.facebook.com/intern/diff/D27410472)

[ghstack-poisoned]
